apisix/discovery/nacos.lua

```diff
@@ -0,0 +1,316 @@
+    end
+
+    if str_byte(url, #url) ~= str_byte("/") then
+        url = url .. "/"
+    end
+
+    return url, username, password
+end
+
+
+local function get_page_service(infos, base_uri, token_param, page_num)
+    local path = str_format(service_list_path, page_num, 100) .. token_param
+    local data, err = get_url(base_uri, path)
+    if err then
+        return data, err, path
+    end
+
+    for _, service_name in ipairs(data.doms) do
+        core.table.insert(infos, service_name)
+    end
+    return data, err, path
+end
+
+
+local function iter_and_add_service_info(infos, base_uri, token_param)
+    local data, err, path = get_page_service(infos, base_uri, token_param, 1)
+    if err then
+        log.error("get_url:" .. path .. " err:" .. err)
+        return
+    end
+
+    local maxPage = math.ceil(data.count / 100)
+    if maxPage == 0 then
+        return
+    end
+
+    if maxPage > 1 then
+        for i = 2, maxPage do
+            get_page_service(infos, base_uri, token_param, i)
+        end
+    end
+end
+
+
+local function get_services(base_uri, token_param)
+    local infos = core.table.new(0, 0)
+    iter_and_add_service_info(infos, base_uri, token_param)
+    return infos
+end
+
+
+local function fetch_full_registry(premature)
+    if premature then
+        return
+    end
+
+    local up_apps = core.table.new(0, 0)
+    local base_uri, username, password = get_base_uri()
+    local token_param, err = get_token_param(base_uri, username, password)
+    if err then
+        if not applications then
+            applications = up_apps
+        end
+        return
+    end
+
+    local infos = get_services(base_uri, token_param)
+    if #infos == 0 then
+        applications = up_apps
+        return
+    end
+
+    local data, err
+    for _, service_name in ipairs(infos) do
+        data, err = get_url(base_uri, instance_list_path .. service_name .. token_param)
+        if err then
+            log.error("get_url:" .. instance_list_path .. " err:" .. err)
+            if not applications then
+                applications = up_apps
+            end
+            return
+        end
+
+        for _, host in ipairs(data.hosts) do
+            if tostring(host.valid) == 'true' and
+                    tostring(host.healthy) == 'true' and
+                    tostring(host.enabled) == 'true' then
+                local nodes = up_apps[service_name]
+                if not nodes then
+                    nodes = core.table.new(0, 0)
+                    up_apps[service_name] = nodes
+                end
+                core.table.insert(nodes, {
+                    host = host.ip,
+                    port = host.port,
+                    weight = host.weight or default_weight,
+                })
+            end
+        end
+    end
+    applications = up_apps
+end
+
+
+function _M.nodes(service_name)
+    while( not applications )
+    do
+        log.info('wait init')
```

Comment:
Better to use a `WARN` level log and only log for the first sleep.
`info` level is invisible by default.

Reply:
OK.
BTW,when i run test file in local CentOS7,most time success,sometime will throw error:
 [error] 23302#23302: *10 [lua] init.lua:462: http_access_phase(): failed to set upstream: no valid upstream node: nil, client: 127.0.0.1, server: localhost, request: "GET /hello HTTP/1.1", host: "localhost"
Is it a bug?